File: Boggle.hs
module Boggle (boggle) where
-- sanna balouchi
import Data.List (isPrefixOf)
import qualified Data.Map as Map
type Board = [[Char]]
type Visited = [[Bool]]

{--
    Fill in the boggle function below. Use as many helpers as you want.
    Test your code by running 'cabal test' from the tester_hs_simple directory.
--}

boggle :: Board -> [String] -> [(String, [(Int, Int)])]
boggle board words =
  let n = length board
      visitedWords = replicate n (replicate n False) -- creates nxn list of visited letters on board
      found = Map.empty -- creates the empty map for words found
  in Map.toList $ solve board words n visitedWords found -- collects all found words

-- takes in the board, list of valid words, size of board, visited list, words found and returns filled map of found words
solve :: Board -> [String] -> Int -> Visited -> Map.Map String [(Int, Int)] -> Map.Map String [(Int, Int)]
solve board words size visited found =
  foldl (\acc1 i ->
           foldl (\acc2 j ->
                    let newVisted = updateVisited i j visited --updates visited
                        current = [board !! i !! j] --grabs letter off board
                    in (dfssearch current i j words acc2 board [(i,j)] newVisted) --continues to dfssearch all possible paths starting with letter
                 ) acc1 [0..size-1]
        ) found [0..size-1]

--conducts dfs for all paths possible
dfssearch :: String -> Int -> Int -> [String] -> Map.Map String [(Int, Int)] -> Board -> [(Int, Int)] -> Visited -> Map.Map String [(Int, Int)]
dfssearch current x y words found board position visited =
  let neighbours = [[1, 0], [0, 1], [-1, 0], [0, -1], [1, 1], [-1, -1], [1, -1], [-1, 1]] -- up, down, left, right, corners
      validWord = filter (isPrefixOf current) words -- returns true if the current word can form a word
      newVisited = updateVisited x y visited    -- updates the visited
      wordsFound = if current `elem` words --if it is a valid word
                   then Map.insert current position found -- then add to map
                   else found -- otherwise remain unchanged
  in if length current > (length board * length board) || null validWord
  then wordsFound --if cannot form a word, or out of bounds, exit dfs
  else --else continue with dfs
    foldl (\wordsacc [dx, dy] ->
        let addx = x + dx --new positions for x,y
            addy = y + dy
        --makes sure it is on the board and not out of bounds + not already visited
        in if addx >= 0 && addx < length board && addy >= 0 && addy < length board && not (newVisited !! addx !! addy) 
           then let newPosition = position ++ [(addx, addy)] --new position list updated to include current new position
                in dfssearch (current ++ [board !! addx !! addy]) addx addy validWord wordsacc board newPosition newVisited -- continues dfs
           else wordsacc --returns found words to move to next dfs
            ) wordsFound neighbours 
  
-- sets the position given to true so we do not visit again
updateVisited :: Int -> Int -> Visited -> Visited
updateVisited i j visited = map (\(idx, row) -> 
    if idx == i 
    then take j row ++ [True] ++ drop (j + 1) row
    else row) 
    (zip [0..] visited)

File: tester_ex_simple/lib/boggle.ex
defmodule Boggle do
  @moduledoc """
    Add your boggle function below. You may add additional helper functions if you desire. 
    Test your code by running 'mix test' from the tester_ex_simple directory.
  """

  #sanna banana 501154179

  def boggle(board, words) do

    word_set = MapSet.new(words) #creates a set of words
    trie = buildTrie(words) #builds prefix tree

    # dfs searches through each letter in the board starting w [0][0]
    Enum.reduce(0..(tuple_size(board)-1), %{}, fn x, acc -> #accumulates the empty map overtime
      Enum.reduce(0..(tuple_size(board)-1), acc, fn y, acc2 ->
        Map.merge(acc2, dfssearch(board, word_set, trie, {x, y}, "", %{}, [])) #merges the maps accumulated
      end)
    end)
  end

  #does a depth first search of the current letter, calls upon the function again using nextElement helper function (grabs the neighbouring letters)
  defp dfssearch(board, word_set, trie, {x, y}, currentWord, visited, position) do
    #makes sure x and y are within the board boundaries, else return empty map, AND THAT IT HAS NOT BEEN VISITED
    if x >= 0 and x < tuple_size(board) and y >= 0 and y < tuple_size(board) and not Map.has_key?(visited, {x, y}) do
      newWord = currentWord <> elem(elem(board, x), y) #concatenates the letter to the previously built word at [x,y]
      newPosition = [{x, y} | position] #adds the current position to the word builder coordinates thing
      newVisited = Map.put(visited, {x, y}, true) #marks the current position as visited

      if Map.has_key?(trie, newWord) do
        if newWord in word_set do #checks if word created so far is in the dictionary
          #puts the word in the map with its coordinates of letters (reversed for proper list of coords)
          Map.put(nextElement(board, word_set, trie, {x, y}, newWord, newVisited, newPosition), newWord, Enum.reverse(newPosition))
        else #otherwise add new letters
          nextElement(board, word_set, trie, {x, y}, newWord, newVisited, newPosition)
        end
      else
        %{} #if not in prefix tree, can instantly return the empty map as there are no words with the letters
      end
    else
      %{} #if was out of board bounds / already visited
    end
  end

  #checks neighbours in the corners, up, down, left, right, done similar to how we initiated dfs
  defp nextElement(board, word_set, trie, {x, y}, currentWord, visited, position) do
    Enum.reduce([-1, 0, 1], %{}, fn addx, acc -> #changes the row number: up, same, down
      Enum.reduce([-1, 0, 1], acc, fn addy, acc2 -> #changes the column number: left, same, right
        Map.merge(acc2, dfssearch(board, word_set, trie, {x+addx, y+addy}, currentWord, visited, position)) #diverges dfs into each neighbour
      end) end)
  end

  #creates a prefix tree, used to let the solver know that the current letters can build a word
  defp buildTrie(words) do
    Enum.reduce(words, %{}, fn word, trie -> #for each word, begin an empty map
      Enum.reduce(0..(String.length(word) - 1), trie, fn i, trie -> #to iterate through the letters
        prefix = String.slice(word, 0..i) #slices the words from its prefixes, ex) tree -> t, tr, tre, etc
        Map.put(trie, prefix, true) #puts each prefix within the prefix tree map
      end) end)
  end
end